File: 2462. Total Cost to Hire K Workers/main.py
import heapq
from typing import List


class Solution:
    def totalCost(self, costs: List[int], k: int, candidates: int) -> int:
        heap1 = []
        heap2 = []
        left = 0
        right = len(costs) - 1

        while left <= right and len(heap1) < candidates:
            heapq.heappush(heap1, costs[left])
            left += 1

        while left <= right and len(heap2) < candidates:
            heapq.heappush(heap2, costs[right])
            right -= 1

        total = 0

        while k > 0:
            if not heap2 or (heap1 and heap1[0] <= heap2[0]):
                total += heapq.heappop(heap1)
                if left <= right:
                    heapq.heappush(heap1, costs[left])
                    left += 1
            elif not heap1 or (heap2 and heap1[0] > heap2[0]):
                total += heapq.heappop(heap2)
                if left <= right:
                    heapq.heappush(heap2, costs[right])
                    right -= 1

            k -= 1

        return total

File: 2462. Total Cost to Hire K Workers/main.rs
struct Solution;

use std::cmp::Reverse;
use std::collections::BinaryHeap;

impl Solution {
    pub fn total_cost(costs: Vec<i32>, k: i32, candidates: i32) -> i64 {
        let mut heap1 = BinaryHeap::new();
        let mut heap2 = BinaryHeap::new();
        let mut left = 0;
        let mut right = costs.len() - 1;

        while left <= right && heap1.len() < candidates as usize {
            heap1.push(Reverse(costs[left]));
            left += 1;
        }

        while left <= right && heap2.len() < candidates as usize {
            heap2.push(Reverse(costs[right]));
            right -= 1;
        }

        let mut total = 0;
        let mut k = k;

        while k > 0 {
            let pop_heap1 = match (heap1.peek(), heap2.peek()) {
                (Some(&Reverse(val1)), Some(&Reverse(val2))) => val1 <= val2,
                (Some(_), None) => true,
                (None, Some(_)) => false,
                (None, None) => break,
            };

            if pop_heap1 {
                if let Some(Reverse(value)) = heap1.pop() {
                    total += value as i64;
                }

                if left <= right {
                    heap1.push(Reverse(costs[left]));
                    left += 1;
                }
            } else {
                if let Some(Reverse(value)) = heap2.pop() {
                    total += value as i64;
                }
                if left <= right {
                    heap2.push(Reverse(costs[right]));
                    right -= 1;
                }
            }

            k -= 1;
        }

        total
    }
}
